networkx/algorithms/hierarchy_coordinates.py

```diff
@@ -0,0 +1,1139 @@
+    produced through maximal layer removal.
+
+    The original algorithm was given in "On the origins of hierarchy in complex networks."[1]_ and with details
+    given in their appendix.[2]_
+
+    Parameters
+    ----------
+    DAG: NetworkX Graph
+        A Directed, Acyclic Graph.
+
+    Return
+    -------
+    float:
+        feedforwardness, hierarchy coordinate
+
+    Examples
+    ---------
+    >>> import numpy as np
+    >>> a = np.array([
+    ...     [0, 0.2, 0, 0, 0],
+    ...     [0.4, 0, 0, 0.7, 0],
+    ...     [0, 0.4, 0, 0, 0],
+    ...     [0, 0, 0.1, 0, 1.0],
+    ...     [0, 0, 0, 0, 0],
+    ... ])
+    >>> condensed_networks, base_binary_networks = nx.node_weighted_condense(a)
+    >>> for network in condensed_networks:
+    ...     print("DAG: ", network)
+    ...     print("Feedforwardness: {0}".format(nx.feedforwardness(network)))
+    DAG:  DiGraph with 2 nodes and 1 edges
+    Feedforwardness: 0.4
+    DAG:  DiGraph with 4 nodes and 3 edges
+    Feedforwardness: 0.7388888888888889
+    DAG:  DiGraph with 5 nodes and 4 edges
+    Feedforwardness: 1.0
+    DAG:  DiGraph with 5 nodes and 4 edges
+    Feedforwardness: 1.0
+    DAG:  DiGraph with 3 nodes and 2 edges
+    Feedforwardness: 1.0
+    DAG:  DiGraph with 3 nodes and 2 edges
+    Feedforwardness: 1.0
+    DAG:  DiGraph with 2 nodes and 1 edges
+    Feedforwardness: 1.0
+    DAG:  DiGraph with 2 nodes and 1 edges
+    Feedforwardness: 1.0
+
+
+    Notes
+    ______
+    feedforwardness is calculated by pruning layers of the original graph and averaging
+    over subsequent feedforwardness calculations.
+
+    TODO: Not sure what's proper re: optional num_thresholds and threshold_distribution parameter
+    .. [1] "On the origins of hierarchy in complex networks."
+     Corominas-Murtra, Bernat, Joaquín Goñi, Ricard V. Solé, and Carlos Rodríguez-Caso,
+     Proceedings of the National Academy of Sciences 110, no. 33 (2013)
+    .. [2] Appendix of 2013 paper:
+    https://www.pnas.org/highwire/filestream/613265/field_highwire_adjunct_files/0/sapp.pdf
+
+    See Also
+    --------
+    treeness, orderability, hierarchy_coordinates
+    """
+    # Must be fed the set of graphs with nodes lower on the hierarchy eliminated first
+    if not nx.is_directed_acyclic_graph(DAG):
+        raise nx.NetworkXError(
+            "DAG must be a directed acyclic graph for feedforwardness evaluation"
+        )
+
+    if nx.get_node_attributes(DAG, "weight") == {}:
+        raise nx.NetworkXError(
+            "Ensure feedforwardness is fed nodes weighted by consdensation"
+            ", e.g. G = nx.weight_nodes_by_condensation(nx.condensation(G))"
+        )
+
+    successively_peeled_nx_graphs = recursive_leaf_removal(DAG, from_top=False)
+    if (
+        len(successively_peeled_nx_graphs) == 1
+        and len(successively_peeled_nx_graphs[0].nodes()) == 1
+    ):
+        return 0
```

Comment:
Does this special case treatment avoid an error later? If not, this could be removed...
If so, then don't we need to check all peeled graphs for a single node?

Reply:
So long as `recursive_leaf_removal` is passed DAGs (with `keep_linkless_layer=False` implicitly), the only time a singleton graph will be passed is when the original DAG itself was a single node, and thus the trivial 0 feedforwardness. I think this occurs not infrequently for well connected graphs whose giant component condenses to a single node, and so this 0 feedforwardness weighting is needed in the final average.

Reply:
So long as `recursive_leaf_removal` is passed DAGs (with `keep_linkless_layer=False` implicitly), the only time a singleton graph will be passed is when the original DAG itself was a single node, and thus the trivial 0 feedforwardness. I think this occurs not infrequently for well connected graphs whose giant component condenses to a single node, and so this 0 feedforwardness weighting is needed in the final average.

Reply:
Then maybe reduce this (and maybe put before the recursive_leaf_removal) to:   ???
```python
if len(DAG) == 1:
    return 0
```

Reply:
done!